### DISCUSSION TITLE
How to use MongoDB Panache simplified update query with multiple properties

### DISCUSSION THREAD

**Discussion by mholzer85** [quarkusio/quarkus#26395] [Q&A] [answered]
I'm using MongoDB Panache with the repository pattern and I'm trying to do an **update setting multiple properties of a document** with a single query.

According to the MongoDB with Panache docs (e.g. [https://quarkus.io/version/2.2/guides/mongodb-panache#simplified-queries](https://quarkus.io/version/2.2/guides/mongodb-panache#simplified-queries) ) it's possible to do so with a simplified query:

> MongoDB with Panache offers operations to update multiple documents based on an update document and a query : Person.update("foo = ?1, bar = ?2", fooName, barName).where("name = ?1", name).

What I've tried doing is this (excerpt of my actual code):
```Java
return update("state = ?1, participantId = ?2, lastUpdate = ?3",
		AppointmentState.BOOKED, participantDto.getId(), dateTimeProvider.currentSystemTime())
		.where("id = ?1 AND state = ?2 AND participantId is null",
				new ObjectId(appointmentId), AppointmentState.FREE);
```

But what this generates as real query against MongoDB is this:
```
2022-06-28 08:26:01,261 DEBUG [org.mon.dri.pro.command] (main) Sending command '{"update": "Appointment", "ordered": true, "writeConcern": {"j": true}, "$db": "rag", "$clusterTime": {"clusterTime": {"$timestamp": {"t": 1656404761, "i": 2}}, "signature": {"hash": {"$binary": {"base64": "AAAAAAAAAAAAAAAAAAAAAAAAAAA=", "subType": "00"}}, "keyId": 0}}, "lsid": {"id": {"$binary": {"base64": "ULbEhCSQRNSOzLqC57zHTg==", "subType": "04"}}}, "updates": [{"q": {"id": {"$oid": "512e1f77bcf86cd799439011"}, "state": "FREE", "participantId": {"$exists": false}}, "u": {"$set": {"state": "BOOKED"}}, "multi": true}]}' with request id 9 to database rag on connection [connectionId{localValue:5, serverValue:7}] to server localhost:55063
```
Notice this part: `"u": {"$set": {"state": "BOOKED"}}`
So **only the first property is set** (`state`), but the others, `participantId` and `lastUpdate`, are completely ignored.

The only thing that worked for me is to not use a simplified (update) query but instead write the actual MongoDB json query for the update:
```Java
return update("{'$set' : {'state': ?1, 'participantId': ?2, 'lastUpdate': ?3}}",
		AppointmentState.BOOKED, participantDto.getId(), dateTimeProvider.currentSystemTime())
		.where("_id = ?1 AND state = ?2 AND participantId is null",
				new ObjectId(appointmentId), AppointmentState.FREE);
```

Which generates something like `"u": {"$set": {"state": "BOOKED", "participantId": 5, "lastUpdate": {"$date": "2022-06-23T12:30:00Z"}}}`.

Did I do something wrong with the simplified query version or is this a bug?

I've tried using Quarkus 2.0.3-Final and 2.2.3-Final. At least for now, I'm not able to upgrade to a more recent version due to other project dependencies.

If that's an actual bug (and I'm not just missing something obvious), I would be happy to provide a pull request with possible solution. 
But someone would have to point me to the correct location / class(es) where the transformation from simplified query to JSON query is done.

[a comment by an automated account is not shown]

**Comment by loicmathieu** [comment, marked as answer]
You should use `and` and not `,` in the update part:

```
return update("state = ?1 and participantId = ?2 and lastUpdate = ?3",
		AppointmentState.BOOKED, participantDto.getId(), dateTimeProvider.currentSystemTime())
		.where("id = ?1 AND state = ?2 AND participantId is null",
				new ObjectId(appointmentId), AppointmentState.FREE);
```

It's documented inside the JavaDoc that specify the update part as a Panache query, but the guide is incorrect about it https://quarkus.io/guides/mongodb-panache#simplified-queries so there is a bug in the guide.

**Reply by mholzer85** [reply]
Thanks for the quick reply. This works as expected.
I created a pull request for updating the guide: https://github.com/quarkusio/quarkus/pull/26407
